src/server/clients_info.rs

```diff
@@ -7,65 +9,77 @@ use bevy_renet::renet::ClientId;
-    /// Stored to reuse allocated capacity.
-    pub(super) entity_buffer: Vec<Vec<Entity>>,
+    /// Stored to reuse allocated memory.
+    info: Vec<ClientInfo>,
 
-    /// Disconnected client's [`ClientsInfo`].
+    /// [`Vec`]'s from acknowledged update indexes from [`ClientInfo`].
     ///
-    /// [`ClientInfo::clear`] is used before the insertion.
-    /// Stored to reuse allocated memory.
-    info_buffer: Vec<ClientInfo>,
+    /// Stored to reuse allocated capacity.
+    entities: Vec<Vec<Entity>>,
 }
 
 impl ClientsInfo {
+    /// Returns a mutable iterator over clients information.
+    pub(super) fn iter_mut(&mut self) -> impl Iterator<Item = &mut ClientInfo> {
+        self.0.iter_mut()
+    }
+
+    /// Returns number of connected clients.
+    pub(super) fn len(&self) -> usize {
+        self.0.len()
+    }
+
     /// Initializes a new [`ClientInfo`] for this client.
-    pub(super) fn init(&mut self, client_id: ClientId) {
-        let client_info = if let Some(mut client_info) = self.info_buffer.pop() {
-            client_info.id = client_id;
+    ///
+    /// Reuses the memory from the buffers if available.
+    pub(super) fn init(&mut self, client_buffers: &mut ClientBuffers, client_id: ClientId) {
+        let client_info = if let Some(mut client_info) = client_buffers.info.pop() {
+            client_info.reset(client_id);
             client_info
         } else {
             ClientInfo::new(client_id)
         };
 
-        self.info.push(client_info);
+        self.0.push(client_info);
     }
 
     /// Removes info for the client.
     ///
-    /// Keeps allocated memory.
-    pub(super) fn remove(&mut self, client_id: ClientId) {
+    /// Keeps allocated memory in the buffers for reuse.
+    pub(super) fn remove(&mut self, client_buffers: &mut ClientBuffers, client_id: ClientId) {
         let index = self
-            .info
+            .0
             .iter()
             .position(|info| info.id == client_id)
             .expect("clients info should contain all connected clients");
-        let mut client_info = self.info.remove(index);
-        self.entity_buffer.extend(client_info.reset());
-        self.info_buffer.push(client_info);
+        let mut client_info = self.0.remove(index);
+        client_buffers.entities.extend(client_info.drain_entities());
+        client_buffers.info.push(client_info);
     }
 
     /// Clears information for all clients.
     ///
-    /// Keeps allocated memory.
-    pub(super) fn clear(&mut self) {
-        for mut client_info in self.info.drain(..) {
-            self.entity_buffer.extend(client_info.reset());
-            self.info_buffer.push(client_info);
+    /// Keeps allocated memory in the buffers for reuse.
+    pub(super) fn clear(&mut self, client_buffers: &mut ClientBuffers) {
+        for mut client_info in self.0.drain(..) {
+            client_buffers.entities.extend(client_info.drain_entities());
+            client_buffers.info.push(client_info);
```

Comment:
I no longer reset it here. Now I do it before the usage instead. It's because I don't know the new `client_id` in advance.

Previously `reset` just didn't reset `id`, and I set this field manually on usage. I found this API a bit weird and switched to this strategy. Plus we use the same approach for our replication buffers for the same reason.